### PR TITLE
Upgrade to credentials-2.0

### PR DESCRIPTION
**Pull request by stephenc**
@reviewbybees

Fix the credentials-2.0 style of store actions and other misc tidy-up for the ew credentials 2.0 api

### CODE REVIEW

File: src/main/java/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider.java [hunk 1]
```diff
@@ -69,17 +61,38 @@
...
 
+    /**
+     * The valid scopes for this store.
+     */
     private static final Set<CredentialsScope> SCOPES =
             Collections.<CredentialsScope>singleton(CredentialsScope.GLOBAL);
 
+    @GuardedBy("self")
+    private static final WeakHashMap<AbstractFolder<?>,FolderCredentialsProperty> emptyProperties =
+            new WeakHashMap<AbstractFolder<?>, FolderCredentialsProperty>();
```
Comment (jglick) [inline src/main/java/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider.java]: IIUC this is designed to solve a problem analogous to the proprietary CJP-1758, correct? To wit: merely saving a folder configuration should not add an empty `FolderCredentialsProperty` to the `config.xml`; this should only be created if and when some credentials are actually added to it.

File: src/main/java/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider.java [hunk 2]
```diff
@@ -424,88 +476,73 @@ private synchronized boolean updateCredentials(@NonNull Domain domain, @NonNull
...
-                } else {
-                    return Collections.emptySet();
-                }
-            }
-
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public AbstractFolderProperty<?> reconfigure(StaplerRequest req, JSONObject form) throws FormException {
```
Comment (jglick) [inline src/main/java/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider.java]: Why?

File: src/main/resources/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider/help.groovy [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    text(_("Credentials will be available to: "));
+    ul() {
+        li() {
+            text(_("Authentication: "))
+            code() {
+                text(ACL.SYSTEM.name);
+            }
+        }
+        def p = CredentialsProvider.USE_ITEM;
+        while (false && p != null && !p.enabled) {
```
Comment (jglick) [inline src/main/resources/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider/help.groovy]: ??

File: src/main/resources/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider/help.groovy [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+                text(_("Users with permission: "))
+                code() {
+                    text(p.group.title);
+                    text("/");
+                    text(p.name);
+                }
+            }
+        }
+    }
+    text(_("Credentials will be available all items within the folder"));
```
Comment (jglick) [inline src/main/resources/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider/help.groovy]: :ant: what?

**Comment by jglick** [comment]
:bee: to the extent I can understand this, but there is no context or explanation for what you are doing here.

File: src/main/resources/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider/help.groovy (hunk 4 above)
Comment (stephenc) [inline src/main/resources/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider/help.groovy]: On the new Manage Jenkins | Configure Credentials | Providers config options, when configuring the enabled providers, this help text will be available.

The help text is generated by Groovy as we need to check which permission is available. If the admin has not enabled Credentials/UseItem then it should follow Item/Configure

File: src/main/resources/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider/help.groovy (hunk 3 above)
Comment (stephenc) [inline src/main/resources/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider/help.groovy]: Good catch, debug text left in play, I shall remove the `false &&`

File: src/main/java/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider.java (hunk 2 above)
Comment (stephenc) [inline src/main/java/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider.java]: To keep the property. If we don't do this then the property will be removed on a form resale as there is no form values in the config.jelly

File: src/main/java/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider.java (hunk 1 above)
Comment (stephenc) [inline src/main/java/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider.java]: Yes, though this solution will save the empty property in some circumstances.

File: src/main/resources/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider/help.groovy (hunk 4 above)
Comment (jglick) [inline src/main/resources/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider/help.groovy]: I was noting that the sentence is ungrammatical.

**Comment by jglick** [comment]
re:bee:

**Pull request merged**